kythe/typescript/indexer.ts

```diff
@@ -1127,6 +1130,48 @@ class Visitor {
     }
   }
 
+  /**
+   * Visits a module declaration, which can look like any of the following:
+   *     declare module 'foo';
+   *     declare module 'foo' {}
+   *     declare module foo {}
+   *     namespace Foo {}
+   */
+  visitModuleDeclaration(decl: ts.ModuleDeclaration) {
+    let sym = this.getSymbolAtLocation(decl.name);
+    if (!sym) {
+      this.todo(
+          decl.name, `module declaration ${decl.name.getText()} has no symbol`);
+      return;
+    }
+    // A TypeScript module declaration declares both a namespace (a Kythe
+    // "record") and a value (most similar to a "package", which defines a
+    // module with declarations).
+    const kNamespace = this.getSymbolName(sym, TSNamespace.NAMESPACE);
+    this.emitNode(kNamespace, 'record');
+    this.emitFact(kNamespace, 'subkind', 'namespace');
+    const kValue = this.getSymbolName(sym, TSNamespace.VALUE);
+    this.emitNode(kValue, 'package');
+
+    const nameAnchor = this.newAnchor(decl.name);
+    this.emitEdge(nameAnchor, 'defines/binding', kNamespace);
+    this.emitEdge(nameAnchor, 'defines/binding', kValue);
+
+    // The entire module declaration defines the created namespace.
+    const defAnchor = this.newAnchor(decl);
+    this.emitEdge(defAnchor, 'defines', kValue);
+
+    if (decl.decorators) this.visitDecorators(decl.decorators);
+    if (decl.body) {
+      this.emitFact(kNamespace, 'complete', 'definition');
```

Comment:
I'm not sure if it matters, but note that a given namespace can be defined (extended) multiple times, e.g.

namespace x { ... }
namespace x { ... more stuff ... }